File: python/src/oxa/openai/responses/decode.py
"""OpenAI Responses face → IR conversions (spec/11 §4)."""

from __future__ import annotations

from typing import Any, cast

from oxa.ir import (
    LOSS_UNMAPPED_FIELD,
    LOSS_UNMAPPED_VALUE,
    LOSS_UNSUPPORTED_SEMANTIC,
    STOP_END_TURN,
    STOP_MAX_TOKENS,
    STOP_OTHER,
    STOP_TOOL_USE,
    Block,
    InputTokensDetails,
    Loss,
    OutputTokensDetails,
    Params,
    ReasoningEffort,
    Request,
    Response,
    TextBlock,
    ThinkingBlock,
    Tool,
    ToolUseBlock,
    Usage,
)
from oxa.ir import (
    ROLE_USER as IR_ROLE_USER,
)
from oxa.ir import (
    Message as IrMessage,
)
from oxa.modelmap import Table
from oxa.openai.responses.constants import (
    INCOMPLETE_REASON_MAX_OUTPUT_TOKENS,
    ITEM_TYPE_FUNCTION_CALL,
    ITEM_TYPE_FUNCTION_CALL_OUTPUT,
    ITEM_TYPE_MESSAGE,
    ITEM_TYPE_REASONING,
    PART_TYPE_OUTPUT_TEXT,
    ROLE_ASSISTANT,
    ROLE_SYSTEM,
    ROLE_USER,
    STATUS_COMPLETED,
    STATUS_FAILED,
    STATUS_INCOMPLETE,
    TOOL_TYPE_FUNCTION,
)
from oxa.openai.responses.normalize import (
    content_system,
    decode_assistant_run,
    decode_content,
    decode_output_run,
    decode_tool_choice,
    loss,
)


def decode_request(
    wire: dict[str, Any],
    raw_text: str = "",
    table: Table | None = None,
) -> tuple[Request, list[Loss]]:
    """Converts a Responses wire request to the IR (face → IR)."""
    losses: list[Loss] = []

    text_config = wire.get("text")
    unmapped_fields = [
        (
            "metadata",
            "metadata",
            wire.get("metadata") is not None,
            "Responses request metadata has no IR equivalent in v1.",
        ),
        (
            "text.verbosity",
            "verbosity",
            isinstance(text_config, dict) and text_config.get("verbosity") is not None,
            "Responses output verbosity has no IR equivalent in v1.",
        ),
        (
            "text.format",
            "format",
            isinstance(text_config, dict) and text_config.get("format") is not None,
            "Responses text output format has no IR equivalent in v1.",
        ),
        (
            "parallel_tool_calls",
            "parallel_tool_calls",
            wire.get("parallel_tool_calls") is not None,
            "Responses parallel tool calls have no IR equivalent in v1.",
        ),
    ]
    for path, field, present, detail in unmapped_fields:
        if present:
            losses.append(loss(path, field, LOSS_UNMAPPED_FIELD, detail))

    model = str(wire.get("model", ""))
    if table is not None:
        model = table.map(model)

    req = Request(
        model=model,
        messages=[],
        system=[],
    )

    if wire.get("instructions"):
        req.system.append(TextBlock(text=wire["instructions"]))

    tools: list[Tool] = []
    for index, tool in enumerate(wire.get("tools", [])):
        kind = tool.get("type", "")
        if kind != TOOL_TYPE_FUNCTION:
            losses.append(
                loss(
                    f"tools[{index}]",
                    "type",
                    LOSS_UNSUPPORTED_SEMANTIC,
                    f"Responses tool type {kind!r} has no IR equivalent",
                )
            )
            continue
        desc = tool.get("description")
        tools.append(
            Tool(
                name=tool.get("name", ""),
                description=desc if desc else None,
                input_schema=tool.get("parameters", {}),
            )
        )
        if tool.get("strict") is not None:
            losses.append(
                loss(
                    f"tools[{index}].strict",
                    "strict",
                    LOSS_UNMAPPED_FIELD,
                    "Responses function tool strict mode has no IR equivalent in v1.",
                )
            )

    tool_choice, tc_losses = decode_tool_choice(wire.get("tool_choice"))
    losses.extend(tc_losses)

    raw_input = wire.get("input")
    if isinstance(raw_input, str):
        req.messages.append(IrMessage(role=IR_ROLE_USER, content=[TextBlock(text=raw_input)]))
    elif isinstance(raw_input, list):
        index = 0
        while index < len(raw_input):
            item = raw_input[index]
            kind = item.get("type", "")
            if kind in ("", ITEM_TYPE_MESSAGE):
                role = item.get("role", "")
                if role == ROLE_SYSTEM:
                    path = f"input[{index}].content"
                    content, c_losses = decode_content(item.get("content"), path)
                    sys_blocks, s_losses = content_system(content, path)
                    req.system.extend(sys_blocks)
                    losses.extend(c_losses)
                    losses.extend(s_losses)
                    index += 1
                elif role == ROLE_USER:
                    path = f"input[{index}].content"
                    content, c_losses = decode_content(item.get("content"), path)
                    losses.extend(c_losses)
                    if not content:
                        content = [TextBlock(text="")]
                    req.messages.append(IrMessage(role=IR_ROLE_USER, content=content))
                    index += 1
                elif role == ROLE_ASSISTANT:
                    msg, next_idx, run_losses = decode_assistant_run(raw_input, index)
                    if msg is not None:
                        req.messages.append(msg)
                    losses.extend(run_losses)
                    index = next_idx
                else:
                    raise ValueError(f"responses: input[{index}]: unknown role {role!r}")
            elif kind in (ITEM_TYPE_FUNCTION_CALL, ITEM_TYPE_REASONING):
                msg, next_idx, run_losses = decode_assistant_run(raw_input, index)
                if msg is not None:
                    req.messages.append(msg)
                losses.extend(run_losses)
                index = next_idx
            elif kind == ITEM_TYPE_FUNCTION_CALL_OUTPUT:
                msg, next_idx = decode_output_run(raw_input, index)
                req.messages.append(msg)
                index = next_idx
            else:
                losses.append(
                    loss(
                        f"input[{index}]",
                        "type",
                        LOSS_UNSUPPORTED_SEMANTIC,
                        f"Responses input item type {kind!r} has no IR equivalent",
                    )
                )
                index += 1

    if not req.messages:
        raise ValueError("responses: request carries no conversation input")

    reasoning_effort: ReasoningEffort | None = None
    reasoning = wire.get("reasoning")
    raw_effort: str | None = None
    if isinstance(reasoning, str):
        raw_effort = reasoning
    elif isinstance(reasoning, dict):
        if "summary" in reasoning:
            losses.append(
                loss(
                    "reasoning.summary",
                    "summary",
                    LOSS_UNMAPPED_FIELD,
                    "Responses reasoning summary preference has no IR equivalent",
                )
            )
        for key in reasoning.keys() - {"effort", "summary"}:
            losses.append(
                loss(
                    f"reasoning.{key}",
                    key,
                    LOSS_UNMAPPED_FIELD,
                    f"Responses reasoning property {key!r} has no IR equivalent",
                )
            )
        effort_value = reasoning.get("effort")
        if effort_value is not None and not isinstance(effort_value, str):
            raise ValueError("responses: reasoning.effort must be a string")
        raw_effort = effort_value
    elif reasoning is not None:
        raise ValueError("responses: reasoning must be an object or string")

    if raw_effort is not None:
        if raw_effort in ("minimal", "low", "medium", "high"):
            reasoning_effort = cast(ReasoningEffort, raw_effort)
        else:
            losses.append(
                loss(
                    "reasoning.effort",
                    "effort",
                    LOSS_UNMAPPED_VALUE,
                    f"unknown reasoning effort value {raw_effort!r}",
                )
            )

    max_tokens = wire.get("max_output_tokens")
    params: Params | None = None
    if (
        wire.get("temperature") is not None
        or wire.get("top_p") is not None
        or max_tokens is not None
        or reasoning_effort is not None
    ):
        params = Params(
            temperature=wire.get("temperature"),
            top_p=wire.get("top_p"),
            max_tokens=max_tokens,
            reasoning_effort=reasoning_effort,
        )

    return (
        Request(
            model=model,
            messages=req.messages,
            system=req.system,
            tools=tools if tools else None,
            tool_choice=tool_choice,
            params=params,
        ),
        losses,
    )


def decode_response(
    wire: dict[str, Any],
    raw_text: str = "",
    table: Table | None = None,
) -> tuple[Response, list[Loss]]:
    """Converts a Responses wire response to the IR (face → IR)."""
    losses: list[Loss] = []
    thinkings: list[Block] = []
    text: list[Block] = []
    calls: list[Block] = []
    has_tool_use = False

    for item_index, item in enumerate(wire.get("output", [])):
        kind = item.get("type", "")
        if kind == ITEM_TYPE_MESSAGE:
            for content_index, part in enumerate(item.get("content", [])):
                p_kind = part.get("type", "")
                if p_kind != PART_TYPE_OUTPUT_TEXT:
                    losses.append(
                        loss(
                            f"output[{item_index}].content[{content_index}]",
                            "type",
                            LOSS_UNSUPPORTED_SEMANTIC,
                            f"Responses output content type {p_kind!r} has no IR equivalent",
                        )
                    )
                    continue
                if part.get("annotations"):
                    losses.append(
                        loss(
                            f"output[{item_index}].content[{content_index}].annotations",
                            "annotations",
                            LOSS_UNMAPPED_FIELD,
                            "Responses output annotations have no IR equivalent in v1.",
                        )
                    )
                text.append(TextBlock(text=part.get("text", "")))
        elif kind == ITEM_TYPE_REASONING:
            summary = item.get("summary", [])
            if not summary:
                losses.append(
                    loss(
                        f"output[{item_index}]",
                        "type",
                        LOSS_UNSUPPORTED_SEMANTIC,
                        "Responses reasoning output item with empty summary carries no convertible content",
                    )
                )
            else:
                for part_index, part in enumerate(summary):
                    if part.get("type") == PART_TYPE_OUTPUT_TEXT:
                        thinkings.append(ThinkingBlock(thinking=part.get("text", "")))
                    else:
                        part_type = part.get("type", "")
                        losses.append(
                            loss(
                                f"output[{item_index}].summary[{part_index}]",
                                "type",
                                LOSS_UNSUPPORTED_SEMANTIC,
                                f"Responses reasoning summary type {part_type!r} has no IR equivalent",
                            )
                        )
            if item.get("encrypted_content"):
                losses.append(
                    loss(
                        f"output[{item_index}].encrypted_content",
                        "encrypted_content",
                        LOSS_UNMAPPED_FIELD,
                        "Responses reasoning encrypted_content has no IR equivalent in v2.",
                    )
                )
        elif kind == ITEM_TYPE_FUNCTION_CALL:
            calls.append(
                ToolUseBlock(
                    id=item.get("call_id", ""),
                    name=item.get("name", ""),
                    input=item.get("arguments", ""),
                )
            )
            has_tool_use = True
        else:
            losses.append(
                loss(
                    f"output[{item_index}]",
                    "type",
                    LOSS_UNSUPPORTED_SEMANTIC,
                    f"Responses output item type {kind!r} has no IR equivalent",
                )
            )

    stop_reason, status_losses = decode_status(wire, has_tool_use)
    losses.extend(status_losses)
    content = thinkings + text + calls

    model = str(wire.get("model", ""))
    if table is not None:
        model = table.map(model)

    usage_raw = wire.get("usage", {})
    input_details = usage_raw.get("input_token_details")
    output_details = usage_raw.get("output_token_details")
    usage = Usage(
        input_tokens=int(usage_raw.get("input_tokens", 0)),
        output_tokens=int(usage_raw.get("output_tokens", 0)),
        input_tokens_details=(
            InputTokensDetails(cached_tokens=int(input_details["cached_tokens"]))
            if isinstance(input_details, dict) and input_details.get("cached_tokens") is not None
            else None
        ),
        output_tokens_details=(
            OutputTokensDetails(reasoning_tokens=int(output_details["reasoning_tokens"]))
            if isinstance(output_details, dict)
            and output_details.get("reasoning_tokens") is not None
            else None
        ),
    )

    return (
        Response(
            id=str(wire.get("id", "")),
            model=model,
            content=content,
            stop_reason=stop_reason,
            usage=usage,
        ),
        losses,
    )


def decode_status(wire: dict[str, Any], has_tool_use: bool) -> tuple[str, list[Loss]]:
    if wire.get("error"):
        error = wire["error"]
        code = error.get("code")
        msg = error.get("message", "")
        return (
            STOP_OTHER,
            [
                loss(
                    "error",
                    "error",
                    LOSS_UNSUPPORTED_SEMANTIC,
                    f"failed Responses response carries error {code!r}: {msg}",
                )
            ],
        )

    status = wire.get("status", "")
    if status == STATUS_COMPLETED:
        return (STOP_TOOL_USE if has_tool_use else STOP_END_TURN), []
    if status == STATUS_INCOMPLETE:
        details = wire.get("incomplete_details", {})
        reason = details.get("reason", "")
        if reason == INCOMPLETE_REASON_MAX_OUTPUT_TOKENS:
            return STOP_MAX_TOKENS, []
        return (
            STOP_OTHER,
            [
                loss(
                    "incomplete_details.reason",
                    "reason",
                    LOSS_UNMAPPED_VALUE,
                    f"Responses incomplete_details reason {reason!r} has no IR equivalent",
                )
            ],
        )
    if status == STATUS_FAILED:
        return (
            STOP_OTHER,
            [
                loss(
                    "error",
                    "error",
                    LOSS_UNSUPPORTED_SEMANTIC,
                    "failed Responses response carries no error object",
                )
            ],
        )
    return STOP_OTHER, []

File: rust/crates/oxa-responses/src/decode.rs
//! Face → IR conversions (spec/11 §4).

use oxa_ir::{
    Block, InputTokensDetails, Loss, LossReason, OutputTokensDetails, Params, ReasoningEffort,
    Request as IrRequest, Response as IrResponse, Role, StopReason, Usage,
};

use crate::config::Config;
use crate::error::Error;
use crate::normalize::{
    content_system, decode_assistant_run, decode_content, decode_output_run, decode_tool_choice,
    loss,
};
use crate::types::{
    INCOMPLETE_REASON_MAX_OUTPUT_TOKENS, ITEM_TYPE_FUNCTION_CALL, ITEM_TYPE_FUNCTION_CALL_OUTPUT,
    ITEM_TYPE_MESSAGE, ITEM_TYPE_REASONING, Input, PART_TYPE_OUTPUT_TEXT, ROLE_ASSISTANT,
    ROLE_SYSTEM, ROLE_USER, Request, Response, STATUS_COMPLETED, STATUS_FAILED, STATUS_INCOMPLETE,
    TOOL_TYPE_FUNCTION, UsageWire,
};

/// Converts a Responses wire request to the IR (face → IR). Semantic
/// unmappables are losses, never errors; errors are reserved for structural
/// type violations of known fields (spec/02 §4).
pub fn decode_request(wire: &Request, config: &Config) -> Result<(IrRequest, Vec<Loss>), Error> {
    let mut losses = Vec::new();
    for (path, field, present, detail) in [
        (
            "metadata",
            "metadata",
            wire.metadata
                .as_ref()
                .is_some_and(|metadata| !metadata.is_empty()),
            "Responses request metadata has no IR equivalent in v1.",
        ),
        (
            "text.verbosity",
            "verbosity",
            wire.text
                .as_ref()
                .and_then(|text| text.verbosity.as_ref())
                .is_some(),
            "Responses output verbosity has no IR equivalent in v1.",
        ),
        (
            "text.format",
            "format",
            wire.text
                .as_ref()
                .and_then(|text| text.format.as_ref())
                .is_some(),
            "Responses text output format has no IR equivalent in v1.",
        ),
        (
            "parallel_tool_calls",
            "parallel_tool_calls",
            wire.parallel_tool_calls.is_some(),
            "Responses parallel tool calls have no IR equivalent in v1.",
        ),
    ] {
        if present {
            losses.push(loss(path, field, LossReason::UnmappedField, detail));
        }
    }

    let mut request = IrRequest {
        model: config.map_model(&wire.model),
        system: Vec::new(),
        messages: Vec::new(),
        tools: None,
        tool_choice: None,
        params: None,
        metadata: None,
    };
    if let Some(instructions) = &wire.instructions {
        request.system.push(oxa_ir::SystemBlock::Text {
            text: instructions.clone(),
        });
    }

    let mut tools = Vec::new();
    for (index, tool) in wire.tools.iter().enumerate() {
        if tool.kind != TOOL_TYPE_FUNCTION {
            losses.push(loss(
                format!("tools[{index}]"),
                "type",
                LossReason::UnsupportedSemantic,
                format!("Responses tool type {:?} has no IR equivalent", tool.kind),
            ));
            continue;
        }
        tools.push(oxa_ir::Tool {
            name: tool.name.clone(),
            description: non_empty(tool.description.clone()),
            input_schema: tool.parameters.clone(),
        });
        if tool.strict.is_some() {
            losses.push(loss(
                format!("tools[{index}].strict"),
                "strict",
                LossReason::UnmappedField,
                "Responses function tool strict mode has no IR equivalent in v1.",
            ));
        }
    }
    if !tools.is_empty() {
        request.tools = Some(tools);
    }
    let (choice, choice_losses) = decode_tool_choice(wire.tool_choice.as_ref());
    request.tool_choice = choice;
    losses.extend(choice_losses);

    match &wire.input {
        Input::Text(text) => request.messages.push(oxa_ir::Message {
            role: Role::User,
            content: vec![Block::Text { text: text.clone() }],
        }),
        Input::Items(items) => {
            let mut index = 0usize;
            while index < items.len() {
                let item = &items[index];
                match item.kind.as_str() {
                    "" | ITEM_TYPE_MESSAGE => match item.role.as_str() {
                        ROLE_SYSTEM => {
                            let path = format!("input[{index}].content");
                            let (content, content_losses) =
                                decode_content(item.content.as_ref(), &path);
                            let (system, system_losses) = content_system(content, &path);
                            request.system.extend(system);
                            losses.extend(content_losses);
                            losses.extend(system_losses);
                            index += 1;
                        }
                        ROLE_USER => {
                            let (mut content, content_losses) = decode_content(
                                item.content.as_ref(),
                                &format!("input[{index}].content"),
                            );
                            if content.is_empty() {
                                content.push(Block::Text {
                                    text: String::new(),
                                });
                            }
                            request.messages.push(oxa_ir::Message {
                                role: Role::User,
                                content,
                            });
                            losses.extend(content_losses);
                            index += 1;
                        }
                        ROLE_ASSISTANT => {
                            let (message, next, run_losses) = decode_assistant_run(items, index)?;
                            if let Some(message) = message {
                                request.messages.push(message);
                            }
                            losses.extend(run_losses);
                            index = next;
                        }
                        other => {
                            return Err(Error::new(format!(
                                "responses: input[{index}]: unknown role {other:?}"
                            )));
                        }
                    },
                    ITEM_TYPE_FUNCTION_CALL | ITEM_TYPE_REASONING => {
                        let (message, next, run_losses) = decode_assistant_run(items, index)?;
                        if let Some(message) = message {
                            request.messages.push(message);
                        }
                        losses.extend(run_losses);
                        index = next;
                    }
                    ITEM_TYPE_FUNCTION_CALL_OUTPUT => {
                        let (message, next) = decode_output_run(items, index);
                        request.messages.push(message);
                        index = next;
                    }
                    other => {
                        losses.push(loss(
                            format!("input[{index}]"),
                            "type",
                            LossReason::UnsupportedSemantic,
                            format!("Responses input item type {other:?} has no IR equivalent"),
                        ));
                        index += 1;
                    }
                }
            }
        }
    }
    if request.messages.is_empty() {
        return Err(Error::new(
            "responses: request carries no conversation input",
        ));
    }
    let reasoning_effort = decode_reasoning_effort(wire.reasoning.as_ref(), &mut losses)?;
    let params = Params {
        temperature: wire.temperature,
        top_p: wire.top_p,
        max_tokens: wire.max_output_tokens,
        stop_sequences: None,
        reasoning_effort,
    };
    if params.temperature.is_some()
        || params.top_p.is_some()
        || params.max_tokens.is_some()
        || params.reasoning_effort.is_some()
    {
        request.params = Some(params);
    }
    Ok((request, losses))
}

fn decode_reasoning_effort(
    value: Option<&serde_json::Value>,
    losses: &mut Vec<Loss>,
) -> Result<Option<ReasoningEffort>, Error> {
    let effort = match value {
        None | Some(serde_json::Value::Null) => return Ok(None),
        Some(serde_json::Value::String(effort)) => effort.as_str(),
        Some(serde_json::Value::Object(reasoning)) => {
            if reasoning.contains_key("summary") {
                losses.push(loss(
                    "reasoning.summary",
                    "summary",
                    LossReason::UnmappedField,
                    "Responses reasoning summary preference has no IR equivalent",
                ));
            }
            match reasoning.get("effort") {
                None | Some(serde_json::Value::Null) => return Ok(None),
                Some(serde_json::Value::String(effort)) => effort.as_str(),
                Some(_) => return Err(Error::new("responses: reasoning.effort must be a string")),
            }
        }
        Some(_) => {
            return Err(Error::new(
                "responses: reasoning must be an object or string",
            ));
        }
    };
    let mapped = match effort {
        "minimal" => Some(ReasoningEffort::Minimal),
        "low" => Some(ReasoningEffort::Low),
        "medium" => Some(ReasoningEffort::Medium),
        "high" => Some(ReasoningEffort::High),
        other => {
            losses.push(loss(
                "reasoning.effort",
                "effort",
                LossReason::UnmappedValue,
                format!("unknown reasoning effort value {other:?}"),
            ));
            None
        }
    };
    Ok(mapped)
}

fn non_empty(value: String) -> Option<String> {
    if value.is_empty() { None } else { Some(value) }
}

/// Converts a Responses wire response to the IR (face → IR). Envelope fields
/// (object, item ids/status) are exempt from losses; usage.total_tokens is
/// derived and therefore ignored.
pub fn decode_response(wire: &Response, config: &Config) -> Result<(IrResponse, Vec<Loss>), Error> {
    let mut losses = Vec::new();
    let mut thinkings = Vec::new();
    let mut text = Vec::new();
    let mut calls = Vec::new();
    let mut has_tool_use = false;
    for (item_index, item) in wire.output.iter().enumerate() {
        match item.kind.as_str() {
            ITEM_TYPE_REASONING => {
                if item.summary.is_empty() {
                    losses.push(loss(
                        format!("output[{item_index}]"),
                        "type",
                        LossReason::UnsupportedSemantic,
                        "Responses reasoning output item with empty summary carries no convertible content",
                    ));
                } else {
                    for (part_index, part) in item.summary.iter().enumerate() {
                        if part.kind == PART_TYPE_OUTPUT_TEXT {
                            thinkings.push(Block::Thinking {
                                thinking: part.text.clone(),
                                signature: None,
                            });
                        } else {
                            losses.push(loss(
                                format!("output[{item_index}].summary[{part_index}]"),
                                "type",
                                LossReason::UnsupportedSemantic,
                                format!(
                                    "Responses reasoning summary type {:?} has no IR equivalent",
                                    part.kind
                                ),
                            ));
                        }
                    }
                }
                if !item.encrypted_content.is_empty() {
                    losses.push(loss(
                        format!("output[{item_index}].encrypted_content"),
                        "encrypted_content",
                        LossReason::UnmappedField,
                        "Responses reasoning encrypted_content has no IR equivalent",
                    ));
                }
            }
            ITEM_TYPE_MESSAGE => {
                for (content_index, part) in item.content.iter().enumerate() {
                    if part.kind != PART_TYPE_OUTPUT_TEXT {
                        losses.push(loss(
                            format!("output[{item_index}].content[{content_index}]"),
                            "type",
                            LossReason::UnsupportedSemantic,
                            format!(
                                "Responses output content type {:?} has no IR equivalent",
                                part.kind
                            ),
                        ));
                        continue;
                    }
                    if !part.annotations.is_empty() {
                        losses.push(loss(
                            format!("output[{item_index}].content[{content_index}].annotations"),
                            "annotations",
                            LossReason::UnmappedField,
                            "Responses output annotations have no IR equivalent in v1.",
                        ));
                    }
                    text.push(Block::Text {
                        text: part.text.clone(),
                    });
                }
            }
            ITEM_TYPE_FUNCTION_CALL => {
                calls.push(Block::ToolUse {
                    id: item.call_id.clone(),
                    name: item.name.clone(),
                    input: item.arguments.clone(),
                });
                has_tool_use = true;
            }
            other => losses.push(loss(
                format!("output[{item_index}]"),
                "type",
                LossReason::UnsupportedSemantic,
                format!("Responses output item type {other:?} has no IR equivalent"),
            )),
        }
    }
    let (stop_reason, status_losses) = decode_status(wire, has_tool_use)?;
    losses.extend(status_losses);
    let mut content = thinkings;
    content.extend(text);
    content.extend(calls);
    Ok((
        IrResponse {
            id: wire.id.clone(),
            model: config.map_model(&wire.model),
            content,
            stop_reason,
            stop_sequence: None,
            usage: decode_usage(wire.usage.as_ref())?,
        },
        losses,
    ))
}

fn decode_usage(wire: Option<&UsageWire>) -> Result<Usage, Error> {
    let Some(wire) = wire else {
        return Ok(Usage::default());
    };
    Ok(Usage {
        input_tokens: nonnegative_usage(wire.input_tokens, "usage.input_tokens")?,
        output_tokens: nonnegative_usage(wire.output_tokens, "usage.output_tokens")?,
        input_tokens_details: wire
            .input_token_details
            .map(|details| {
                nonnegative_usage(
                    details.cached_tokens,
                    "usage.input_token_details.cached_tokens",
                )
                .map(|cached_tokens| InputTokensDetails { cached_tokens })
            })
            .transpose()?,
        output_tokens_details: wire
            .output_token_details
            .map(|details| {
                nonnegative_usage(
                    details.reasoning_tokens,
                    "usage.output_token_details.reasoning_tokens",
                )
                .map(|reasoning_tokens| OutputTokensDetails { reasoning_tokens })
            })
            .transpose()?,
        ..Usage::default()
    })
}

fn nonnegative_usage(value: i64, path: &str) -> Result<i64, Error> {
    if value < 0 {
        return Err(Error::new(format!(
            "responses: {path} must be non-negative"
        )));
    }
    Ok(value)
}

pub(crate) fn decode_status(
    wire: &Response,
    has_tool_use: bool,
) -> Result<(StopReason, Vec<Loss>), Error> {
    if let Some(error) = &wire.error {
        return Ok((
            StopReason::Other,
            vec![loss(
                "error",
                "error",
                LossReason::UnsupportedSemantic,
                format!(
                    "failed Responses response carries error {:?}: {}",
                    error.code, error.message
                ),
            )],
        ));
    }
    match wire.status.as_str() {
        STATUS_COMPLETED => Ok((
            if has_tool_use {
                StopReason::ToolUse
            } else {
                StopReason::EndTurn
            },
            Vec::new(),
        )),
        STATUS_INCOMPLETE => {
            let reason = wire
                .incomplete_details
                .as_ref()
                .map(|details| details.reason.as_str())
                .unwrap_or_default();
            if reason == INCOMPLETE_REASON_MAX_OUTPUT_TOKENS {
                Ok((StopReason::MaxTokens, Vec::new()))
            } else {
                Ok((
                    StopReason::Other,
                    vec![loss(
                        "incomplete_details.reason",
                        "reason",
                        LossReason::UnmappedValue,
                        format!(
                            "Responses incomplete_details reason {reason:?} has no IR equivalent"
                        ),
                    )],
                ))
            }
        }
        STATUS_FAILED => Ok((
            StopReason::Other,
            vec![loss(
                "error",
                "error",
                LossReason::UnsupportedSemantic,
                "failed Responses response carries no error object",
            )],
        )),
        other => Err(Error::new(format!(
            "responses: status {other:?} has no IR equivalent"
        ))),
    }
}
